toplevel/identity/entity.go

```diff
@@ -0,0 +1,642 @@
+							if ok {
+								aliasMap[k] = v
+							}
+						}
+					}
+				}
+			}
+
+			entity.Policies = policies
+			entity.Metadata = make(map[string]interface{})
+			metadataMap, ok := entity.Metadata.(map[string]interface{})
+			if ok {
+				for k, v := range metadata {
+					metadataMap[k] = v
+				}
+			}
+		}(&entities[i])
+	}
+	bwg.Wait()
+}
+
+// Calls vault.DiffItems for existing/desired list of aliases, within each exisitng/desired entity
+func determineAliasActions(entries, existingEntities []entity, entitiesToBeDeleted []vault.Item) (map[string]map[string][]vault.Item,
+	[]vault.Item, map[string][]vault.Item) {
+
+	existingEntityToAliases := make(map[string][]entityAlias)
+	for _, entity := range existingEntities {
+		existingEntityToAliases[entity.Name] = append(existingEntityToAliases[entity.Name], entity.Aliases...)
+	}
+	aliasesToBeWritten := make(map[string]map[string][]vault.Item)
+	aliasesToBeDeleted := make([]vault.Item, 0)
+	aliasesToBeUpdated := make(map[string][]vault.Item)
+
+	for _, entry := range entries {
+		w, d, u := vault.DiffItems(aliasesAsItems(entry.Aliases), aliasesAsItems(existingEntityToAliases[entry.Name]))
+		// new entities will not have an id.. need to differentiate organization for alias to be written
+		// by id for existing entity receiving new alias OR new entity with new aliases
+		if entry.Id == "" {
+			if len(aliasesToBeWritten["name"]) == 0 {
+				aliasesToBeWritten["name"] = make(map[string][]vault.Item)
+			}
+			aliasesToBeWritten["name"][entry.Name] = append(aliasesToBeWritten["name"][entry.Name], w...)
+		} else {
+			if len(aliasesToBeWritten["id"]) == 0 {
+				aliasesToBeWritten["id"] = make(map[string][]vault.Item)
+			}
+			aliasesToBeWritten["id"][entry.Id] = append(aliasesToBeWritten["id"][entry.Id], w...)
+		}
+		aliasesToBeDeleted = append(aliasesToBeDeleted, d...)
+		aliasesToBeUpdated[entry.Id] = append(aliasesToBeUpdated[entry.Id], u...)
+	}
+
+	// the parent existing entity DNE in desired (to be deleted)
+	// treat this as deletion of all affiliated aliases
+	for _, e := range entitiesToBeDeleted {
+		if _, exists := existingEntityToAliases[e.(entity).Name]; exists {
+			aliasesToBeDeleted = append(aliasesToBeDeleted, aliasesAsItems(e.(entity).Aliases)...)
+		}
+	}
+	return aliasesToBeWritten, aliasesToBeDeleted, aliasesToBeUpdated
+}
+
+// writes, deletes, and/or updates entity aliases
+func performAliasReconcile(aliasesToBeWritten map[string]map[string][]vault.Item,
+	aliasesToBeDeleted []vault.Item, aliasesToBeUpdated map[string][]vault.Item, customMetadataSupported bool) error {
+	var accessorIds map[string]string
+	// extra work (vault api request) required to organize accessor ids
+	if len(aliasesToBeWritten) > 0 {
+		accessorIds = make(map[string]string)
+		authBackends := vault.ListAuthBackends()
+		for k, v := range authBackends {
+			accessorIds[strings.TrimRight(k, "/")] = v.Accessor
+		}
+	}
+	if _, exists := aliasesToBeWritten["id"]; exists {
```

Comment:
I want to understand the thought on this block and the next in person

Reply:
Understood. A call will be worthwhile. The logic for aliases became fairly more complex than that for entities due to alias dependency on entity creation beforehand.